artifact/v3/version/version.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+const int supported_parser_version {3};
+const string supported_parser_format {"mender"};
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
```

Comment:
Yes, but like we discussed previously, the very best would be `json::From(io::Reader&)`

Reply:
That will require that function to read a file into a buffer anyway. Or a stream interface to our `Reader` because C++ already has IO utilities -- the streams -- an things are setup to work with them not our Go-like implementation.

Reply:
This is true, it is just a question about the size of the buffer. Like buffering the whole payload, or doing it in chunks.

I actually had a conversation with @kacf about it this morning, where I suggested to move to the stdlib interface, and his argument is that the `io-streams` add too much code bloat for us, as they are all templated.

I'm not quite sure I fully agree, but I do understand the argument, and it is not our biggest thorn in our sides imo.

Therefore for now, going with this intermediary buffering is the way it seems.